util/include/Serializable.h

```diff
@@ -37,72 +40,199 @@ typedef std::unique_ptr<char[], std::default_delete<char[]>> UniquePtrToChar;
+  }
+  template<typename T, typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
+  static void deserialize_impl(std::istream& inStream, T& t, int) {
+    inStream.read((char*)&t, sizeof(T));
+    LOG_TRACE(logger(), t);
+  }
+  /** ****************************************************************************************************************/
+  static void serialize(std::ostream& outStream, const char* p, const std::size_t& size) {
+    outStream.write(p, static_cast<std::streamsize>(size));
+  }
+  static void deserialize(std::istream& inStream, char* p, const std::size_t& size) {
+    inStream.read(p, static_cast<std::streamsize>(size));
+  }
 
-  virtual void serialize(std::ostream &) const final;
-
-  virtual std::string getName() const = 0;
-  virtual std::string getVersion() const = 0;
-
-  static void verifyClassVersion(const std::string &expectedVersion, std::istream &inStream);
-  static SerializablePtr deserialize(const UniquePtrToChar &inBuf,
-                                     int64_t inBufSize); //__attribute__ ((deprecated)) TODO [TK] remove;
-  static SerializablePtr deserialize(std::istream &inStream);
+  static concordlogger::Logger& logger(){
+    static concordlogger::Logger logger_ = concordlogger::Log::getLogger("serializable");
+    return logger_;
+  }
 
- protected:
-  virtual void serializeClassName(std::ostream &) const final;
-  virtual void serializeClassVersion(std::ostream &) const final;
-  virtual void serializeDataMembers(std::ostream &) const = 0;
-  virtual void deserializeDataMembers(std::istream &) = 0;
-  virtual SerializablePtr create(std::istream &) = 0;
+  typedef std::unordered_map<std::string, std::function<Serializable*()>> Registry;
 
-  void serializeString(const std::string &str, std::ostream &outStream) const;
-  template<typename INT>
-  void serializeInt(const INT &num, std::ostream &outStream) const {
-    outStream.write((char *) &num, sizeof(INT));
-  }
-  static std::string deserializeClassName(std::istream &inStream);
-  static std::string deserializeClassVersion(std::istream &inStream);
-  static std::string deserializeString(std::istream &inStream);
-  template<typename INT>
-  static INT deserializeInt(std::istream &inStream) {
-    INT res = 0;
-    inStream.read((char *) &res, sizeof(INT));
-    return res;
+  static Registry& registry(){
+    static Registry registry_;
+    return registry_;
   }
+};
 
-  static void registerObject(const std::string &className, const SerializablePtr &objectPtr) {
-    objectFactory_[className] = objectPtr;
+/** *******************************************************************************************************************
+ *  Class for automatic Serializable object registration
+ */
+template<typename T>
+class SerializableFactory: public virtual Serializable {
+public:
+  SerializableFactory(){(void)registered_;}
+  static bool registerT() {
+    registry()[typeid(T).name()] = []() -> Serializable* { return new T;};
```

Comment:
Is a `typeid` guaranteed to stay the same across compilations and across compiler versions?

Reply:
added name demangling

Reply:
Hmm, I thought name demangling was likely good enough, but this comment is concerning. If we can't guarantee the same type name we can have very hard to find bugs.
 
> Returns an implementation defined null-terminated character string containing the name of the type. No guarantees are given; in particular, the returned string can be identical for several types and change between invocations of the same program.

https://en.cppreference.com/w/cpp/types/type_info/name

Reply:
demangled is just a qualified class name: [namespace::]*ClassName

Reply:
Yep. Agreed. I don't see how this can be a problem in practice. Let's stick with it for now, especially since we'll have an assert in case demangling fails, which we'd see immediately in tests.